src/mpid/ch4/shm/xpmem/xpmem_send.h

```diff
@@ -12,52 +12,101 @@
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_LMT_ISEND);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_LMT_RTS_ISEND);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_LMT_RTS_ISEND);
+
+    MPIDI_Datatype_check_contig_size_lb(datatype, count, is_contig, data_sz, true_lb);
+    /* SHM only issues contig large message through XPMEM.
+     * TODO: support noncontig send message */
+    MPIR_Assert(is_contig && data_sz > 0);
 
     sreq = MPIDIG_request_create(MPIR_REQUEST_KIND__SEND, 2);
     MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
     *request = sreq;
+    MPIDIG_REQUEST(sreq, buffer) = (void *) buf;
+    MPIDIG_REQUEST(sreq, datatype) = datatype;
+    MPIDIG_REQUEST(sreq, count) = count;
+    MPIDIG_REQUEST(sreq, rank) = rank;
+    MPIDIG_REQUEST(sreq, tag) = tag;
+    MPIDIG_REQUEST(sreq, context_id) = comm->context_id + context_offset;
+    MPIDI_XPMEM_REQUEST(sreq, counter) = NULL;
 
-    MPIDI_Datatype_check_contig_size_lb(datatype, count, is_contig, data_sz, true_lb);
+    slmt_rts_hdr->src_offset = (uint64_t) buf + true_lb;
+    slmt_rts_hdr->data_sz = data_sz;
+    slmt_rts_hdr->sreq_ptr = (uint64_t) sreq;
+    slmt_rts_hdr->src_lrank = MPIDI_XPMEM_global.local_rank;
+
+    /* message matching info */
+    slmt_rts_hdr->src_rank = comm->rank;
+    slmt_rts_hdr->tag = tag;
+    slmt_rts_hdr->context_id = comm->context_id + context_offset;
+
+    XPMEM_TRACE("lmt_rts_isend: shm ctrl_id %d, buf_offset 0x%lx, data_sz 0x%lx, "
+                "sreq_ptr 0x%lx, local_rank %d, match info[dest %d, src_rank %d, "
+                "tag %d, context_id 0x%x]\n",
+                MPIDI_SHM_XPMEM_SEND_LMT_RTS, slmt_rts_hdr->src_offset, slmt_rts_hdr->data_sz,
+                slmt_rts_hdr->sreq_ptr, slmt_rts_hdr->src_lrank, rank, slmt_rts_hdr->src_rank,
+                slmt_rts_hdr->tag, slmt_rts_hdr->context_id);
+
+    mpi_errno = MPIDI_SHM_do_ctrl_send(rank, comm, MPIDI_SHM_XPMEM_SEND_LMT_RTS, &ctrl_hdr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_LMT_RTS_ISEND);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
 
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_lmt_cts_isend(const void *buf, MPI_Aint count,
+                                                       MPI_Datatype datatype, int rank,
+                                                       MPIR_Comm * comm, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t data_sz;
+    MPI_Aint true_lb;
+    bool is_contig;
+    MPIDI_SHM_ctrl_hdr_t ctrl_hdr;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_cts_t *slmt_cts_hdr = &ctrl_hdr.xpmem_lmt_cts;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_LMT_CTS_ISEND);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_LMT_CTS_ISEND);
+
+    MPIDI_Datatype_check_contig_size_lb(datatype, count, is_contig, data_sz, true_lb);
     MPIR_Assert(is_contig && data_sz > 0);
 
```

Comment:
I do not think you need check `is_contig` on the receiver side.

Reply:
the current coopeartive implementation only supports the sender and receiver's data both are contiguous. I add it for a simple check, can delete if unnecessary